### PR TITLE
feat: gateway object validation

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

Adds validation for `Gateway` objects in our admission webhook.

**Which issue this PR fixes**

Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1946

**Special notes for your reviewer**:

Blocked on https://github.com/kubernetes-sigs/gateway-api/pull/952

**PR Readiness Checklist**:

- [x] the `CHANGELOG.md` release notes have been updated

### CODE REVIEW

**Comment by shaneutt** [comment]
blocked until https://github.com/kubernetes-sigs/gateway-api/pull/952 is resolved.

File: test/integration/webhook_test.go
Comment (mflendrich) [inline test/integration/webhook_test.go:223]: wouldn't refactoring this into a table-based test help readability and ease addition of further test cases?

**Review by mflendrich — COMMENTED** [review]
testing suggestion - I was thinking about possible further test cases that would be worthwhile, and about ease of adding further test cases; hence the suggestion of the table pattern. Otherwise lgtm

File: test/integration/webhook_test.go
Comment (shaneutt) [reply test/integration/webhook_test.go:223]: Sure, I'll change it to a loop

Comment (shaneutt) [reply test/integration/webhook_test.go:223]: So I'm not actually sure, because I tested the logic paths with a _single_ object in this case so my thinking is making it into a loop where I suddenly have several objects would be a lot more code and ultimately might _reduce_ readability? LMKYT

**Comment by shaneutt** [comment]
https://github.com/kubernetes-sigs/gateway-api/pull/952 is merged, https://github.com/Kong/kubernetes-ingress-controller/pull/2047/commits/29013a10640c7449d049fcaff71191f47be1b182 resolves the block.

File: test/integration/webhook_test.go
Comment (mflendrich) [reply test/integration/webhook_test.go:223]: We now have a test that validates unrelated bits of functionality (secret validation and gateway validation) within a single test that is 700 lines long, has the flows for individual isolatable tests mixed, and has unclear patterns of reuse. I'm worried about maintainability of such a test I'd like to look into that (and maybe the nondeterminism coming from the `sleep`) in a PR to this PR I've just started drafting. Stay tuned :radio:

Comment (mflendrich) [reply test/integration/webhook_test.go:223]: https://github.com/Kong/kubernetes-ingress-controller/pull/2051

> the testing suite is due for a refactor where admission webhook setup is part of environment setup and not part of specific test setup such that multiple tests can separately validate that functionality

#2051 makes it possible for admission webhook setup to happen in the context of a specific test case, which (I think) is better than setting up as part of the environment, because it puts the filtering logic closer to the validating logic, and does not bloat the test harness further. The approach proposed in that PR makes it possible to have multiple tests.

I think that we're at the point where the amount of technical debt accrued in the integration tests is too much. Flakiness is a real problem; I view simplifying (and isolating) test cases to be a good countermeasure.

**Review by mflendrich — COMMENTED** [review]
To me this is good to go after #2051 gets merged into this PR.

File: test/integration/webhook_test.go
Comment (shaneutt) [reply test/integration/webhook_test.go:223]: > because it puts the filtering logic closer to the validating logic, and does not bloat the test harness further.

I don't see globally configuring the admission webhook as bloat, I instead see it as additional verification of functionality. We ask end-users to have it on at all times to catch mistakes and we can just as easily make mistakes in tests that might pass if the webhook were off but still technically be invalid or wrong in some way.

Comment (shaneutt) [reply test/integration/webhook_test.go:223]: > I think that we're at the point where the amount of technical debt accrued in the integration tests is too much

This I agree with

>  Flakiness is a real problem; I view simplifying (and isolating) test cases to be a good countermeasure.

Have you seen specific test flakes occur with this test? While I agree it needs work I haven't actually seen it be flaky for webhook testing specific reasons it appears to be consistently passing.

Comment (shaneutt) [reply test/integration/webhook_test.go:223]: I'll review #2051 and we can reconvene here after that.

Comment (shaneutt) [reply test/integration/webhook_test.go:223]: Alright I'm done with my review of #2051 and I think we can move forward with that. I appreciate the willingness to jump in and make a PR to make the changes you wanted to see.

**Review by mflendrich — DISMISSED** [review]
after #2051, lgtm

In my last pass I've noticed that the webhook uses the wrong `context.Context` - `context.Background()` instead of the `ctx` passed from the caller. This was a (benign) bug in the admission webhook handler before, that was only replicated here. Filing an issue (https://github.com/Kong/kubernetes-ingress-controller/issues/2055) to handle that async.

**Review by mflendrich — APPROVED** [review]

**Pull request merged**
